### PR TITLE
Add a C++ implementation of the PN532TokenProvider

### PR DESCRIPTION
**Pull request by hikinggrass**
Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

### CODE REVIEW

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+}
+
+void auth_token_providerImpl::ready() {
+    serial.run();
+
+    serial.reset();
+    auto configure_sam = serial.configureSAM();
+    if (configure_sam.get()) {
+        EVLOG_debug << "Configured SAM" << std::endl;
```
Comment (a-w50) [inline modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:25]: Maybe a bit more verbose? What is SAM?

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    while (true) {
+        auto in_list_passive_target_response = serial.inListPassiveTarget().get();
+        if (in_list_passive_target_response.valid) {
+            std::shared_ptr<InListPassiveTargetResponse> in_list_passive_target_response_message =
+                std::dynamic_pointer_cast<InListPassiveTargetResponse>(in_list_passive_target_response.message);
+            auto target_data = in_list_passive_target_response_message->target_data;
+            for (auto entry : target_data) {
+                EVLOG_info << "Got raw rfid/nfc tag: " << entry.getNFCID() << std::endl;
+                Object rfid_token;
```
Comment (a-w50) [inline modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:43]: could use the json initializer list notation here

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                rfid_token["token"] = entry.getNFCID();
+                rfid_token["type"] = "rfid";
+                rfid_token["timeout"] = config.timeout;
+                EVLOG_info << "Publishing new rfid/nfc token: " << rfid_token;
+                this->publish_token(rfid_token);
+            }
+        }
+
+        std::this_thread::sleep_for(std::chrono::seconds(config.read_timeout));
```
Comment (a-w50) [inline modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:52]: do we really sleep for seconds?

File: modules/PN532TokenProvider/pn532_serial/evSerial.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2022 Pionix GmbH and Contributors to EVerest
+#ifndef YETI_SERIAL
```
Comment (a-w50) [inline modules/PN532TokenProvider/pn532_serial/evSerial.h:3]: wrong define here

**Review by a-w50 — CHANGES_REQUESTED** [review]
I guess, this could be now refactored, using the evSerial from the framework, right?
Concerning the low-level code, there are quite a lot of constants / magic numbers which could be cleaned up.  I would also advise to use packed structs if possible, to decode the messages.  Is the heavy use of promises, shared ptrs, dynamic casts really necessary?

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp (hunk 3 above)
Comment (hikinggrass) [reply modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:53]: Yes, that way we wont see the same token multiple times when its not immediately removed

**Comment by hikinggrass** [comment]
> I guess, this could be now refactored, using the evSerial from the framework, right? Concerning the low-level code, there are quite a lot of constants / magic numbers which could be cleaned up. I would also advise to use packed structs if possible, to decode the messages. Is the heavy use of promises, shared ptrs, dynamic casts really necessary?

I'd merge it like this and then apply the general serial/date/threads PR afterwards

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp (hunk 3 above)
Comment (a-w50) [reply modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:53]: ah ok, but that's quite hacky, isn't it? We would need some kind of debouncing.  But maybe this simple solution works best

Comment (hikinggrass) [reply modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:53]: It is, but I'll have to look into the spec a little more to see if there's something the module does to help with debouncing

Comment (corneliusclaussen) [reply modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:53]: Most implementations I have seen generate two events: the first when the card is seen and the second when the card leaves the RFID reader. The time between those two events is also sometimes used as a user input: some charging stations can be resetted by holding the card down for more then 10 seconds or so

**Review by a-w50 — APPROVED** [review]

File: modules/PN532TokenProvider/main/auth_token_providerImpl.cpp (hunk 3 above)
Comment (hikinggrass) [reply modules/PN532TokenProvider/main/auth_token_providerImpl.cpp:53]: I just looked into the implementation of the js version we are using at the moment, they also use a polling interval: https://github.com/techniq/node-pn532/blob/master/src/pn532.js#L47

**Pull request merged**